File: frontend/src/lib/memory.ts
import { supabase } from "@/lib/supabase";

export interface ConversationMessage {
  role: "user" | "assistant";
  content: string;
  timestamp: string;
  confidence?: string;
}

export interface ConversationHistory {
  conversation_id: string;
  messages: ConversationMessage[];
  summary: string;
  created_at: string;
  updated_at: string;
}

export async function getConversationHistory(
  repoFullName: string,
): Promise<ConversationHistory | null> {
  const { data, error } = await supabase.auth.getSession();

  if (error || !data.session) {
    throw new Error("User not authenticated");
  }

  const token = data.session.access_token;

  try {
    const res = await fetch(
      `${import.meta.env.VITE_API_URL || "https://repomind-577n.onrender.com"}/memory/conversation?repo_full_name=${encodeURIComponent(repoFullName)}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    );

    if (!res.ok) {
      console.error("Failed to fetch conversation history");
      return null;
    }

    return res.json();
  } catch (err) {
    console.error("Error fetching conversation history:", err);
    return null;
  }
}

export async function clearConversation(
  repoFullName: string,
): Promise<boolean> {
  const { data, error } = await supabase.auth.getSession();

  if (error || !data.session) {
    throw new Error("User not authenticated");
  }

  const token = data.session.access_token;

  try {
    const res = await fetch(
      `${import.meta.env.VITE_API_URL || "https://repomind-577n.onrender.com"}/memory/conversation?repo_full_name=${encodeURIComponent(repoFullName)}`,
      {
        method: "DELETE",
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    );

    return res.ok;
  } catch (err) {
    console.error("Error clearing conversation:", err);
    return false;
  }
}

export async function exportConversation(
  repoFullName: string,
): Promise<string | null> {
  const { data, error } = await supabase.auth.getSession();

  if (error || !data.session) {
    throw new Error("User not authenticated");
  }

  const token = data.session.access_token;

  try {
    const res = await fetch(
      `${import.meta.env.VITE_API_URL || "https://repomind-577n.onrender.com"}/memory/conversation/export?repo_full_name=${encodeURIComponent(repoFullName)}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    );

    if (!res.ok) {
      return null;
    }

    const response = await res.json();
    return response.data;
  } catch (err) {
    console.error("Error exporting conversation:", err);
    return null;
  }
}

File: backend/app/routes/memory.py
from fastapi import APIRouter, Depends, HTTPException, Query
from app.auth.supabase import get_current_user
from app.brain.memory.store import get_memory_store

router = APIRouter(prefix="/memory")

memory_store = get_memory_store()

@router.get("/conversation")
async def get_conversation(
    repo_full_name: str = Query(...),
    user=Depends(get_current_user),
):
    """Get conversation history for user and repo"""
    try:
        conversation = memory_store.get_conversation(user["id"], repo_full_name)
        
        if not conversation:
            return {
                "conversation_id": None,
                "messages": [],
                "summary": "No conversation yet"
            }
        
        return {
            "conversation_id": conversation.conversation_id,
            "messages": [
                {
                    "role": msg.role,
                    "content": msg.content,
                    "timestamp": msg.timestamp.isoformat(),
                    "confidence": msg.confidence
                }
                for msg in conversation.messages
            ],
            "summary": conversation.get_conversation_summary(),
            "created_at": conversation.created_at.isoformat(),
            "updated_at": conversation.updated_at.isoformat()
        }
    except Exception as e:
        print(f"Error fetching conversation: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/conversation")
async def clear_conversation(
    repo_full_name: str = Query(...),
    user=Depends(get_current_user),
):
    """Clear conversation history"""
    try:
        memory_store.clear_conversation(user["id"], repo_full_name)
        
        return {
            "status": "success",
            "message": f"Cleared conversation for {repo_full_name}"
        }
    except Exception as e:
        print(f"Error clearing conversation: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/conversation/export")
async def export_conversation(
    repo_full_name: str = Query(...),
    user=Depends(get_current_user),
):
    """Export conversation as JSON"""
    try:
        data = memory_store.export_conversation(user["id"], repo_full_name)
        
        if not data:
            raise HTTPException(status_code=404, detail="No conversation found")
        
        return {
            "status": "success",
            "data": data
        }
    except Exception as e:
        print(f"Error exporting conversation: {e}")
        raise HTTPException(status_code=500, detail=str(e))